docs/architecture/adr-0023-text-array-object-acrs.md

```diff
@@ -0,0 +1,166 @@
+
+In terms of querying for objects with filters on object level ACRs, postgres array functions such as `array_cat()` and `concat_ws()`, and operators such as `&&` (which determines overlay of two arrays), etc., could potentially be handy.
+
+When listing actions or permissions by subject type and id, which will potentially require going into each string of the text array, we could use postgres array functions such as `unnest()` combined with `like` to match string patterns.
+
+### Migration
+
+Migrations could be a bit tricky and probably could take some time to run. Adding a new field and updating this field in all first-class objects' tables would likely require a novel approach of creating a temp table with the new column, writing to a new permanent table, and dropping the previous table. Further research on it will be in a future issue card.
+
+Note that the `acrs` field will have an empty text array as the default value.
+
+### Examples for datasources
+
+This section shows some initial attempts for adapting to the ACR text array solution. These are for initial tests only and don't represent final implementation details. Datasources could be a simple case, but such adaptations can be applied to other first-class objects in the similar way.
+
+#### Data models
+
+```scala
+@JsonCodec
+case class Datasource(
+  id: UUID,
+  createdAt: java.sql.Timestamp,
+  createdBy: String,
+  modifiedAt: java.sql.Timestamp,
+  modifiedBy: String,
+  owner: String,
+  name: String,
+  visibility: Visibility,
+  composites: Json,
+  extras: Json,
+  bands: Json,
+  licenseName: Option[String],
+  acrs: List[String]
+)
+
+@JsonCodec
+case class ObjectAccessControlRuleCreate (
+  subjectType: SubjectType,
+  subjectIdO: Option[String],
+  actionType: ActionType
+) {
+  val subjectId: String = subjectIdO match {
+    case Some(subjectId) => subjectId
+    case _ => ""
+  }
+
+  def toAccessControlRuleString: String =
+    s"${subjectType.toString};${subjectId};${actionType.toString}"
+}
+```
+
+#### CRUD
+
+```scala
+def getAcrsByObjectId(object: Datasource): Fragment =
+  fr"SELECT acrs FROM datasources where id = ${datasource.id};"
+
+def updatePermissions(objectAcrCreateList: List[ObjectAccessControlRuleCreate], datasourceId: UUID, user: User): ConnectionIO[Int] = {
+    val now = new Timestamp((new java.util.Date()).getTime())
+    val objectAcrs = objectAcrCreateList.map(_.toAccessControlRuleString)
+    val updateQuery =
+      fr"UPDATE" ++ this.tableF ++ fr"SET" ++
+      fr"""
+        modified_at = ${now},
+        modified_by = ${user.id},
+        acrs = ${objectAcrs}
+      where id = ${datasourceId}
+      """
+    updateQuery.update.run
+}
+```
+
+#### First-class objects query filtered by ACRs
+
+```scala
+def exampleListQuery(user: User, actionType: ActionType, tableF: Fragment): Fragment = {
+  val baseF: Fragment = fr"SELECT * FROM" ++ tableF ++ "WHERE"
```

Comment:
I would remove the single letter extensions on these and just spell it out - again for ease of reading and avoiding thinking there are typos, etc.